diesel/src/query_builder/combination_clause.rs

```diff
@@ -0,0 +1,206 @@
+where
+    T: Table,
+    Self: Query,
+{
+    type Values = InsertFromSelect<Self, T::AllColumns>;
+
+    fn values(self) -> Self::Values {
+        InsertFromSelect::new(self)
+    }
+}
+
+/// Extension trait to combine queries using a combinator like `UNION`, `INTERSECT` or `EXPECT`
+/// with or without `ALL` rule for duplicates
+pub trait SelectCombinationQueryMethods: Query + Sized {
+    /// Combine two queries using a SQL `UNION`
+    ///
+    /// # Examples
+    /// ```rust
+    /// # #[macro_use] extern crate diesel;
+    /// # include!("../doctest_setup.rs");
+    /// # use schema::{users, animals};
+    ///
+    /// # fn main() {
+    /// #     use self::users::dsl::{users, name as user_name};
+    /// #     use self::animals::dsl::{animals, name as animal_name};
+    /// #     use diesel::query_builder::combination_clause::SelectCombinationQueryMethods;
+    /// #     let connection = establish_connection();
+    /// let data = users.select(user_name.nullable())
+    ///     .union(animals.select(animal_name))
+    ///     .load(&connection);
+    ///
+    /// let expected_data = vec![
+    ///     None,
+    ///     Some(String::from("Jack")),
+    ///     Some(String::from("Tess")),
+    ///     Some(String::from("Sean")),
+    /// ];
+    /// assert_eq!(Ok(expected_data), data);
+    /// # }
+    /// ```
+    fn union<Rhs>(self, rhs: Rhs) -> Combination<Union, Distinct, Self, Rhs>
+    where
+        Rhs: Query<SqlType = Self::SqlType>,
+    {
+        Combination::new(Union, Distinct, self, rhs)
+    }
+
+    /// Combine two queries using a SQL `UNION ALL`
+    fn union_all<Rhs>(self, rhs: Rhs) -> Combination<Union, All, Self, Rhs>
+    where
+        Rhs: Query<SqlType = Self::SqlType>,
+    {
+        Combination::new(Union, All, self, rhs)
+    }
+
+    /// Combine two queries using a SQL `INTERSECT`
+    fn intersect<Rhs>(self, rhs: Rhs) -> Combination<Intersect, Distinct, Self, Rhs>
+    where
+        Rhs: Query<SqlType = Self::SqlType>,
+    {
+        Combination::new(Intersect, Distinct, self, rhs)
+    }
+
+    /// Combine two queries using a SQL `INTERSECT ALL`
+    fn intersect_all<Rhs>(self, rhs: Rhs) -> Combination<Intersect, All, Self, Rhs>
+    where
+        Rhs: Query<SqlType = Self::SqlType>,
+    {
+        Combination::new(Intersect, All, self, rhs)
+    }
+
+    /// Combine two queries using a SQL `EXCEPT`
+    fn except<Rhs>(self, rhs: Rhs) -> Combination<Except, Distinct, Self, Rhs>
+    where
+        Rhs: Query<SqlType = Self::SqlType>,
+    {
+        Combination::new(Except, Distinct, self, rhs)
+    }
+
+    /// Combine two queries using a SQL `EXCEPT ALL`
+    fn except_all<Rhs>(self, rhs: Rhs) -> Combination<Except, All, Self, Rhs>
+    where
+        Rhs: Query<SqlType = Self::SqlType>,
+    {
+        Combination::new(Except, All, self, rhs)
+    }
+}
+
+impl<T: Query> SelectCombinationQueryMethods for T {}
```

Comment:
We probably want to implement that for `T: Table`, `SelectStatement` and `BoxedSelectStatement` explicitly.